Sources/NIOExtras/HTTP1ProxyConnectHandler.swift

```diff
@@ -0,0 +1,226 @@
+//===----------------------------------------------------------------------===//
+//
+// This source file is part of the SwiftNIO open source project
+//
+// Copyright (c) 2022 Apple Inc. and the SwiftNIO project authors
+// Licensed under Apache License v2.0
+//
+// See LICENSE.txt for license information
+// See CONTRIBUTORS.txt for the list of SwiftNIO project authors
+//
+// SPDX-License-Identifier: Apache-2.0
+//
+//===----------------------------------------------------------------------===//
+
+import NIOCore
+import NIOHTTP1
+
+public final class HTTP1ProxyConnectHandler: ChannelDuplexHandler, RemovableChannelHandler {
+    public typealias OutboundIn = Never
+    public typealias OutboundOut = HTTPClientRequestPart
+    public typealias InboundIn = HTTPClientResponsePart
+
+    enum State {
+        // transitions to `.connectSent` or `.failed`
+        case initialized
+        // transitions to `.headReceived` or `.failed`
+        case connectSent(Scheduled<Void>)
+        // transitions to `.completed` or `.failed`
+        case headReceived(Scheduled<Void>)
+        // final error state
+        case failed(Error)
+        // final success state
+        case completed
+    }
+
+    private var state: State = .initialized
+
+    private let targetHost: String
+    private let targetPort: Int
+    private let headers: HTTPHeaders
+    private let deadline: NIODeadline
+
+    private var proxyEstablishedPromise: EventLoopPromise<Void>?
+    var proxyEstablishedFuture: EventLoopFuture<Void>? {
```

Comment:
Is this just for testing or is it meant to be public API?

Reply:
I think we should let the user pass in a promise instead. This also allows `proxyEstablishedFuture` to be non-nil as we would no longer need to delay creation of the promise until the `ChannelHandler` is added to the pipeline to get hold of the `EventLoop`.

Reply:
Is there any reason we shouldn't use a user inbound event instead if it's not just for testing? (I'm okay keeping a promise/future as well.)

Reply:
It is not just for testing:
https://github.com/swift-server/async-http-client/blob/5bee16a79922e3efcb5cea06ecd27e6f8048b56b/Sources/AsyncHTTPClient/ConnectionPool/HTTPConnectionPool%2BFactory.swift#L211-L221

Reply:
I added an `EventLoopPromise` to the initialiser.